olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp

```diff
@@ -51,13 +52,35 @@ std::string GetDatastoreServerUrl(const std::string& partition) {
              ? "https://api-lookup." + server_url->second + "/lookup/v1"
              : "";
 }
+
+class LookupApiLock {
+ public:
+  std::unique_lock<std::mutex> Lock(const client::HRN& catalog) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    const std::string& key = catalog.ToString();
+    std::unique_lock<std::mutex> catalog_lock(mutexes_[key]);
+    return catalog_lock;
+  }
+
+ private:
+  std::mutex mutex_;
+  std::unordered_map<std::string, std::mutex> mutexes_;
+};
```

Comment:
I would rather create some helper class that is wrapped around a static mutex/unordered_map and automatically locks or waits for lock internally and removes the mutex from the unordered_map when it leaves the scope.